README.md
# JsonParser
JsonParser is simple JSON parsing library, you can parse JSONObject or JSONArray using path and key. Simply you just need to pass the name of the object or object key and index number of the array item in your path.  

## Installation

This JsonParser JAR adds in the `lib` folder then after set in the build path. Download JAR form this link [JsonParser_DLab_1.1.jar](Release/JsonParser_DLab_1.1.jar). For, the more information refer this [`JsonParser`](src/main/) sample project.  

## Path
Path is combination of the object key and index number of the array item.  

`Dpath=[menu/items/12/label]`    

here, `menu`  `items` `label` is your object key and 12 is array item index number now, the question arises how to parse using this path so let's little talk on it. When you give path to the library, it is converted path to token then after, first check is JSONObject or JSONArray if object then token must have key value(if not found then it is throwing exception and process break) then after this process recursively doing till your path's all token finished, once token finished then process break and return the result in the Object format.  

> Note : Array index is starting from the 1 so, your array first item index is 1.

### Example:

You can refer more example [here](src/main/JsonParserTest.java), 

```json
{"menu": {
    "header": "SVG Viewer",
    "items": [
        {"id": "Open"},
        {"id": "OpenNew", "label": "Open New"},
        null,
        {"id": "ZoomIn", "label": "Zoom In"},
        {"id": "ZoomOut", "label": "Zoom Out"},
        {"id": "OriginalView", "label": "Original View"},
        null,
        {"id": "Quality"},
        {"id": "Pause"},
        {"id": "Mute"},
        null,
        {"id": "Find", "label": "Find..."}
    ]
}}
```

```java
new JsonParser().addObserver(new OnValueObserver() {
	@Override
	public void onValueFound(String key, Object value) {
		System.out.println(String.valueOf("Find By Path : " + value.toString()));
	}
	
}).inFind(GetJson.strJson_3, "Dpath=[menu/items/12/label]"); 

// Output :- Find By Path : Find...
```

## Key
Key is a unique element of the object so, duplication of the key is not allowed in the object but whole JSONObject or JSONArray may be have same name key more then one time.

`Dkey=[value]`    

here, `value` is your key and more then one time if it exists in the whole object then during this recursively process, if key found then notify the user on `onValueFound(String key, Object value)` method.

### Example:

You can refer more example [here](src/main/JsonParserTest.java), 

```json
{"menu": {
  "id": "file",
  "value": "File",
  "popup": {
    "menuitem": [
      {"value": "New", "onclick": "CreateNewDoc()"},
      {"value": "Open", "onclick": "OpenDoc()"},
      {"value": "Close", "onclick": "CloseDoc()"}
    ]
  }
}}
```

```java
new JsonParser().addObserver(new OnValueObserver() {
	@Override
	public void onValueFound(String key, Object value) {
		System.out.println(String.valueOf("Find By Key : " + value));
	}
	
}).inFind(GetJson.strJson_4, "Dkey=[value]");

// Output :- 
// Find By Key : New
// Find By Key : Open
// Find By Key : Close
```

***

### End User

* QA Engineer, Developer and who not like to do manually parsing of the JSONObject and JSONArray.
* How I get the idea to develope this type of library, when I am helping to the one of the my QA friend in his automation process that time, I observed that he man facing trouble on the parsing of the JSONObject then I decided to develop this type of library, which have structure as like selenium Xpath.


### Current issues

* Integer key value of object's is not supported
* Only JSONObject or JSONArray as string passing in the arguments
* multiple path is not supported
* array of key or array or path not supported as arguments
* always need listener as a notifier

### Change Log

Refer this file for the library version informations [here](ChangeLog),

### Contributing

Contributions are welcome! If you find a bug please report it and if you want add new feature then please suggest to me. If you want to contribute code please file an issue and create a branch off of the current dev branch and file a pull request.

### About me

Kishan Donga ([@ikishan92](https://twitter.com/ikishan92))  

### License

`JsonParser` is released under the [Apache License 2.0](LICENSE).

ChangeLog
-- Change Log --

18/05/2018
Version = 1.0
=> First release 

24/05/2018
Version = 1.1
=> Json Array Index now change 0 to 1 so, your first array element index is 1 not 0


src/main/GetJson.java
package main;

public class GetJson {
	
	public static String strJson_4 = "{\"menu\": {\r\n" +
			"  \"id\": \"file\",\r\n" +
			"  \"value\": \"File\",\r\n" +
			"  \"popup\": {\r\n" +
			"    \"menuitem\": [\r\n" +
			"      {\"value\": \"New\", \"onclick\": \"CreateNewDoc()\"},\r\n" +
			"      {\"value\": \"Open\", \"onclick\": \"OpenDoc()\"},\r\n" +
			"      {\"value\": \"Close\", \"onclick\": \"CloseDoc()\"}\r\n" +
			"    ]\r\n" +
			"  }\r\n" +
			"}}";
	
	public static String strJson_3 = "{\"menu\": {\r\n" +
			"    \"header\": \"SVG Viewer\",\r\n" +
			"    \"items\": [\r\n" +
			"        {\"id\": \"Open\"},\r\n" +
			"        {\"id\": \"OpenNew\", \"label\": \"Open New\"},\r\n" +
			"        null,\r\n" +
			"        {\"id\": \"ZoomIn\", \"label\": \"Zoom In\"},\r\n" +
			"        {\"id\": \"ZoomOut\", \"label\": \"Zoom Out\"},\r\n" +
			"        {\"id\": \"OriginalView\", \"label\": \"Original View\"},\r\n" +
			"        null,\r\n" +
			"        {\"id\": \"Quality\"},\r\n" +
			"        {\"id\": \"Pause\"},\r\n" +
			"        {\"id\": \"Mute\"},\r\n" +
			"        null,\r\n" +
			"        {\"id\": \"Find\", \"label\": \"Find...\"},\r\n" +
			"        {\"id\": \"FindAgain\", \"label\": \"Find Again\"},\r\n" +
			"        {\"id\": \"Copy\"},\r\n" +
			"        {\"id\": \"CopyAgain\", \"label\": \"Copy Again\"},\r\n" +
			"        {\"id\": \"CopySVG\", \"label\": \"Copy SVG\"},\r\n" +
			"        {\"id\": \"ViewSVG\", \"label\": \"View SVG\"},\r\n" +
			"        {\"id\": \"ViewSource\", \"label\": \"View Source\"},\r\n" +
			"        {\"id\": \"SaveAs\", \"label\": \"Save As\"},\r\n" +
			"        null,\r\n" +
			"        {\"id\": \"Help\"},\r\n" +
			"        {\"id\": \"About\", \"label\": \"About Adobe CVG Viewer...\"}\r\n" +
			"    ]\r\n" +
			"}}";
	
	public static String strJson_2 = "{\r\n" +
			"    \"name\":\"John\",\r\n" +
			"    \"age\":30,\r\n" +
			"    \"cars\": {\r\n" +
			"        \"car1\":\"Ford\",\r\n" +
			"        \"car2\":\"BMW\",\r\n" +
			"        \"car3\":\"Fiat\"\r\n" +
			"    },\r\n" +
			"    \"No\": [\"1\",\"2\",\"3\"]\r\n" +
			" }";

	public static String strJson_1 = "{\r\n" +
			"\t\"problems\": [{\r\n" +
			"\t\t\"Diabetes\": [{\r\n" +
			"\t\t\t\"medications\": [{\r\n" +
			"\t\t\t\t\"medicationsClasses\": [{\r\n" +
			"\t\t\t\t\t\"className\": [{\r\n" +
			"\t\t\t\t\t\t\"associatedDrug\": [{\r\n" +
			"\t\t\t\t\t\t\t\"name\": \"asprin\",\r\n" +
			"\t\t\t\t\t\t\t\"dose\": \"\",\r\n" +
			"\t\t\t\t\t\t\t\"strength\": \"500 mg\"\r\n" +
			"\t\t\t\t\t\t}],\r\n" +
			"\t\t\t\t\t\t\"associatedDrug#2\": [{\r\n" +
			"\t\t\t\t\t\t\t\"name\": \"somethingElse\",\r\n" +
			"\t\t\t\t\t\t\t\"dose\": \"\",\r\n" +
			"\t\t\t\t\t\t\t\"strength\": \"500 mg\"\r\n" +
			"\t\t\t\t\t\t}]\r\n" +
			"\t\t\t\t\t}],\r\n" +
			"\t\t\t\t\t\"className2\": [{\r\n" +
			"\t\t\t\t\t\t\"associatedDrug\": [{\r\n" +
			"\t\t\t\t\t\t\t\"name\": \"asprin\",\r\n" +
			"\t\t\t\t\t\t\t\"dose\": \"\",\r\n" +
			"\t\t\t\t\t\t\t\"strength\": \"500 mg\"\r\n" +
			"\t\t\t\t\t\t}],\r\n" +
			"\t\t\t\t\t\t\"associatedDrug#2\": [{\r\n" +
			"\t\t\t\t\t\t\t\"name\": \"somethingElse\",\r\n" +
			"\t\t\t\t\t\t\t\"dose\": \"\",\r\n" +
			"\t\t\t\t\t\t\t\"strength\": \"500 mg\"\r\n" +
			"\t\t\t\t\t\t}]\r\n" +
			"\t\t\t\t\t}]\r\n" +
			"\t\t\t\t}]\r\n" +
			"\t\t\t}],\r\n" +
			"\t\t\t\"labs\": [{\r\n" +
			"\t\t\t\t\"missing_field\": \"missing_value\"\r\n" +
			"\t\t\t}]\r\n" +
			"\t\t}],\r\n" +
			"\t\t\"Asthma\": [{}]\r\n" +
			"\t}]\r\n" +
			"}";
}


src/dlab/JsonParser.java
package dlab;

import java.util.Iterator;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.json.JSONTokener;

/**
 * Created by Kishan Donga on 18-05-18
 */

public class JsonParser {
	
	private OnValueObserver listener;
	private String path;
	
	public void inFind(String jsonString, String dElements) {
		try
		{
			if(dElements != null && !dElements.isEmpty()){
				if(dElements.contains("Dpath")){
					path = dElements.substring(7, dElements.length() - 1);
					String token[] = path.split("/");
					inFindByPath(jsonString, token);
				
				} else if(dElements.contains("Dkey")) {
					String key = dElements.substring(6, dElements.length() - 1);
					Object tokener = getJSONTokener(jsonString);
					
					if(isJSONObject(tokener)){
						inFind(new JSONObject(jsonString), key);
					}
					
					if(isJSONArray(tokener)) {
						inFind(new JSONArray(jsonString), key);
					}
				} else {
					throw new Exception("Invalid Key or Path");
				}
			} else {
				throw new Exception("Key or Path can't null or empty!");
			}
		}
		catch (Exception e) {
			e.printStackTrace();
		}
	}
	
	private void inFindByPath(String jsonString, String[] token) throws JSONException{
		Object object = null;
		
		Object tokener = getJSONTokener(jsonString);
		
		if(isJSONObject(tokener)){
			object = new JSONObject(jsonString);
		}
		
		if(isJSONArray(tokener)) {
			object = new JSONArray(jsonString);
		}
		
		for(String vt : token){
			if(isInt(vt)){
				object = inFindByPath((JSONArray)object, (Integer.parseInt(vt) - 1));
			}else{
				object = inFindByPath((JSONObject)object, vt);
			}
		}
		
		listener.onValueFound(path, object);
	}
	
	private Object inFindByPath(JSONObject object, String value) throws JSONException{
		return object.get(value);
	}
	
	private Object inFindByPath(JSONArray array, int value) throws JSONException{
		return array.get(value);
	}
	
	@SuppressWarnings("rawtypes")
	private void inFind(JSONObject object, String key) throws JSONException {
		Iterator iterator = getObjectKeys(object);
		while (iterator.hasNext()) {
			String k = (String) iterator.next();
            if(key.equals(k)){
            	listener.onValueFound(key, object.get(k));
            } else {
            	Object type = object.get(k);
            	if(type instanceof JSONObject){
            		inFind((JSONObject)type, key);
            	}
            	else if(type instanceof JSONArray){
            		inFind((JSONArray)type, key);
            	}
            }
        }
	}
	
	public boolean isInt(String value){
		try{
			Integer.parseInt(value);
			return true;
		} catch(Exception e){
			return false;
		}
	}
	
    public void inFind(JSONArray array, String key) throws JSONException {
		int count = array.length();
		for(int i = 0; i < count; i++){
			Object type = array.get(i);
			if(type instanceof JSONObject){
        		inFind((JSONObject)type, key);
        	}
        	else if(type instanceof JSONArray){
        		inFind((JSONArray)type, key);
        	}
		}
	}
	
	public JsonParser addObserver(OnValueObserver lst){
		this.listener = lst;
		return this;
	}

	@SuppressWarnings("rawtypes")
	public Iterator getObjectKeys(JSONObject data) throws JSONException{
		return data.keys();
	}
	
	public boolean isJSONObject(String data) throws JSONException {
		Object json = new JSONTokener(data).nextValue();
		if (json instanceof JSONObject){
			return true;
		}else{
			return false;
		}
	}
	
	public boolean isJSONArray(String data) throws JSONException {
		Object json = new JSONTokener(data).nextValue(); 
		if (json instanceof JSONArray){
			return true;
		}else{
			return false;
		}
	}
	
	public boolean isJSONObject(Object object) throws JSONException {
		if (object instanceof JSONObject){
			return true;
		}else{
			return false;
		}
	}
	
	public boolean isJSONArray(Object object) throws JSONException {
		if (object instanceof JSONArray){
			return true;
		}else{
			return false;
		}
	}
	
	public Object getJSONTokener(String data) throws JSONException {
		return new JSONTokener(data).nextValue();
	}
	
	public interface OnValueObserver {
		void onValueFound(String key, Object object);
	}
}


src/main/JsonParserTest.java
package main;

import dlab.JsonParser;
import dlab.JsonParser.OnValueObserver;

/**
 * Created by Kishan Donga on 18-05-18
 */

public class JsonParserTest {

	public static void main(String[] args) {
		
		new JsonParser().addObserver(new OnValueObserver() {
			@Override
			public void onValueFound(String key, Object value) {
				System.out.println(String.valueOf("Find By Key : " + value));
			}
			
		}).inFind(GetJson.strJson_4, "Dkey=[value]");
		
		// Output :- 
		// Find By Key : New
		// Find By Key : Open
		// Find By Key : Close
		
		new JsonParser().addObserver(new OnValueObserver() {
			@Override
			public void onValueFound(String key, Object value) {
				System.out.println(String.valueOf("Find By Path : " + value.toString()));
			}
			
		}).inFind(GetJson.strJson_3, "Dpath=[menu/items/12/label]"); 
		
		// Output :- Find By Path : Find...
			
		new JsonParser().addObserver(new OnValueObserver() {
			@Override
			public void onValueFound(String key, Object value) {
				System.out.println(String.valueOf("Find By Path : " + value.toString()));
			}
			
		}).inFind(GetJson.strJson_2, "Dpath=[No/1]"); 
		
		// Output :- Find By Path : 1
		
		new JsonParser().addObserver(new OnValueObserver() {
			@Override
			public void onValueFound(String key, Object value) {
				
				new JsonParser().addObserver(new OnValueObserver() {
					@Override
					public void onValueFound(String key, Object value) {
						System.out.println(String.valueOf("Find By Path : " + value.toString()));
					}
					
				}).inFind(value.toString(), "Dpath=[medicationsClasses/1/className/1/associatedDrug/1/name]"); 
			}
			
		}).inFind(GetJson.strJson_1, "Dpath=[problems/1/Diabetes/1/medications/1/]"); 
		
		// Output :- Find By Path : asprin
		
		new JsonParser().addObserver(new OnValueObserver() {
			@Override
			public void onValueFound(String key, Object value) {
				System.out.println(String.valueOf("Find By Key : " + value));
			}
			
		}).inFind(GetJson.strJson_1, "Dkey=[name]");
		
		// Output :- 
		// Find By Key : asprin
		// Find By Key : somethingElse
		// Find By Key : asprin
		// Find By Key : somethingElse
	}
}
